### ISSUE TITLE
Test & bugfix scrapers

### ISSUE THREAD

**Issue by robredpath** [open-contracting/kingfisher#100]
- [ ] armenia.py
- [ ] australia.py
- [x] canada_buyandsell.py
- [x] canada_montreal.py
- [x] colombia.py
- [x] mexico_administracion_publica_federal.py
- [x] mexico_cdmx.py
- [x] mexico_grupo_aeroporto.py #122 
- [x] mexico_inai.py
- [x] mexico_jalisco.py
- [ ] moldova.py
- [ ] paraguay_dncp.py
- [ ] paraguay_hacienda.py
- [x] taiwan.py
- [ ] uganda.py
- [ ] uk_contracts_finder.py #103
- [ ] ukraine.py #102
- [x] uruguay.py

Tick off above when checked or when broken out to a separate ticket (if we realise one is having big problems let's make it a separate ticket and try and keep this one relatively clean - they will probably be separate pull requests anyway!)

**Comment by robredpath** [comment]
It's worth noting that the server we have for this has 12 cores and AFAIK at the very least the gather and fetch stages are independent, so we should be able to run quite of lot of these in parallel

**Comment by odscjames** [comment]
#108 should contain a useful function in the DB class we can reuse but here's my first quick try at some SQL .... get the ID of the source_session row you want to delete. In one transaction run:

````
delete from release_check where release_id in (select release.id from release join source_session_file_status on source_session_file_status.id =  release.source_session_file_status_id where source_session_file_status.source_session_id = $id);

delete from record_check where record_id in (select record.id from record join source_session_file_status on source_session_file_status.id =  record.source_session_file_status_id where source_session_file_status.source_session_id = $id);

delete from record where source_session_file_status_id in (select id from source_session_file_status where source_session_id = $id);

delete from release where source_session_file_status_id in (select id from source_session_file_status where source_session_id = $id);

delete from source_session_file_status where source_session_id = $id;

delete from source_session where id = $id;
````

Then in one transaction per statement run:

````
delete from data where id not in (select data_id from release union select data_id from record);

delete from package_data where id not in (select package_data_id from release union select package_data_id from record);
````

Not sure how this will work at scale yet but we'll start by assuming the DB is better than what we could do and only look at optimizations later if we need to. We'll properly test and play as we do other work.

**Comment by odscjames** [comment]
Have started a bunch by hand, most are in TMUX sessions:

> taiwan
> colombia
> canada_montreal
> canada_buyandsell
> mexico_administracion_publica_federal
> mexico_cdmx
> mexico_grupo_aeroporto
> mexico_inai
> uruguay

Picked these ones because the number of files in each one is known in advance, and we know it's not insane (~4000 files is I think the biggest one). We'll check in later.

**Comment by robredpath** [comment]
@tian2992 's up for getting involved with this as well

**Comment by jpmckinney** [comment]
Just checking whether this branch can be deleted (it seems very similar code for a Columbia scraper is now part of the repo): https://github.com/open-contracting/ocdsdata/compare/colombia

**Comment by odscjames** [comment]
@jpmckinney  Deleted

**Comment by odscjames** [comment]
This is progressing well, the first batch has already shook out some problems that have all been either P.R.'d or commented on in other issues.

For now, I'm going to tick them off if the source can run a full run, including the new stage of "check". Maybe some hand analysis of the data should then be done to make sure it's sensible - that may uncover issues like the Ukraine issue, where it turned out we were downloading way more than we had to. But how to do that is unclear at this stage, and others have more skill than me at that.

**Comment by tian2992** [comment]
Have tested out colombia, however I'm kinda concerned on the default download strategy as it is quite a significant load to extract several GB of data.. Deserves its own, separate discussion.

**Comment by odscjames** [comment]
> Have tested out colombia, however I'm kinda concerned on the default download strategy as it is quite a significant load to extract several GB of data.. Deserves its own, separate discussion.

Thanks! One question is, is there any sensible way that a user might want to say "only download X of Colmbia"? For instance, see #95 - it will be possible to only download a historical subset of Ukraine’s data in the future. Is there any easy variable that Colmbia's data can be divided by? If so, do make a separate issue and lets put that in!

**Comment by odscjames** [comment]
BTW I am currently running Armenia and Australia on the ETL server. Australia is still going and Armenia has died with an error, so I can look into that now.

**Comment by tian2992** [comment]
Currently for colombia's data there is either bulk or --sample, which covers only the first page: about 10k entries.

**Comment by tian2992** [comment]
Ran mexico_administracion_publica_federal with no issues, now checking the resulted inputs on DB

**Comment by jpmckinney** [comment]
For Moldova, regarding the code comment about invalid JSON, you just need to replace the last trailing comma. It's better to do that than hardcode (and then have to update the script every year), e.g.:

```python
import json
import requests

response = requests.get('http://opencontracting.date.gov.md/multiple-file-api/releases.json')
data = json.loads(''.join(response.text.rsplit(',', 1)))

for url in data['links']['all']:
    # more code
```

**Comment by jpmckinney** [comment]
For Uganda, I had noticed when scraping it in December 2017 that if the JSON contained a `status_code` (in which case, it also contained a `message`), then the response didn't contain data. We should check for this, and not emit data for that page if present.

I notice that the present scraper has a different filename pattern for samples and non-samples.

**Comment by jpmckinney** [comment]
Added comments to new issues. Closing here.
